### DISCUSSION TITLE
How to update a notify element with new content / arguments?

### DISCUSSION THREAD

**Discussion by pythonormen** [zauberzeug/nicegui#515] [Ideas / Feature Requests]
Could someone provide a simple example on how to access/modify an existing notification with new arguments or content similar to this Quasar example?

I've tried but it's just generating new notifications, lost on how to access the existing one.

Thanks

https://quasar.dev/quasar-plugins/notify#updatable-notifications

<img width="607" alt="image" src="https://user-images.githubusercontent.com/125542430/224919448-168b1e3e-4835-4191-b175-7cf46ea23871.png">

**Comment by falkoschindler** [comment]
That would be a useful feature!
Unfortunately, this is currently not possible and I'm struggling to find a workaround.
We will probably need to re-think the `ui.notify` API.

**Reply by abel1502** [reply]
Why not make ui.notify an object, like most other ui elements? It would act as a handle through which the notification will be manipulateable. Something similar to the dialog api should probably be applicable

**Reply by falkoschindler** [reply]
@abel1502 In my point of view the difference between dialogs and notifications is that dialogs are usually reused throughout the lifetime of an application, while notifications are usually fired once and forgotten. If there is a notification every minute and the application runs for a year (imagine home automation) that's half a million notifications. If we keep every notification like a regular UI element that's quite some memory usage. And if we even keep a representation in the DOM, the initial page load would be pretty slow if there are already plenty of old notifications around.
But maybe some kind of hybrid is possible: Keep references to notifications that are still visible and make sure to prune unused references.

**Reply by thaije** [reply]
I had a similar issue in #1994. Perhaps a simple solution is to add an argument to the notification UI element that makes them persistent and into objects, e.g: 
```python
btn = ui.button('Say hi!', on_click=lambda: ui.notify('Hi!', persistent=True close_button='OK'))
```

This way by default and for most applications notifications are memory efficient and fire and forget, but if more functionality is required it is available by setting the persistent argument. 
Persistent notifications would then allow for changing the actual notification object, such as changing its contents, or adding a button with an onclick listener similar to #1994 .

**Reply by falkoschindler** [reply]
I just played around with `Quasar.Notify` and tried to create an updatable notification. Apparently one needs to create it with something like
```js
n = Quasar.Notify.create({message: 'Hi!', group: false})
```
Then we can update it with
```js
n({message: 'Ho!'})
```

So I'm wondering if we should move control over notifications to the server, i.e.
- make `ui.notify` a proper NiceGUI element,
- which creates a new Quasar notification (timeout=0) when mounted,
- and which closes the notification and deletes itself after a given timeout (if the notification isn't persistent).

This would give us much more control over the notification! 😀

**Reply by falkoschindler** [reply]
I created PR #2031 to play around with such a notification element. Still work in progress.

**Reply by falkoschindler** [reply]
The PR has been merged and will be released shortly. I'll close this discussion.
